helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -813,23 +817,36 @@ public void onStateChange(String instanceName, List<CurrentState> statesInfo,
 
   @Override
   @PreFetch(enabled = false)
-  public void onCustomizedStateRootChange(String instanceName, NotificationContext changeContext) {
+  public void onCustomizedStateRootChange(String instanceName, List<String> customizedStateTypes,
+      NotificationContext changeContext) {
     logger.info("START: GenericClusterController.onCustomizedStateRootChange()");
-    notifyCaches(changeContext, ChangeType.CUSTOMIZED_STATE_ROOT);
     HelixManager manager = changeContext.getManager();
-    List<String> customizedStateTypes =
-        manager.getHelixDataAccessor().getChildNames(
-        manager.getHelixDataAccessor().keyBuilder().customizedStatesRoot(instanceName));
+    Builder keyBuilder = new Builder(manager.getClusterName());
 
-    for (String customizedState : customizedStateTypes) {
-      try {
-        manager.addCustomizedStateChangeListener(this, instanceName, customizedState);
-        logger.info(
-            manager.getInstanceName() + " added customized state listener for " + instanceName
-                + ", listener: " + this);
-      } catch (Exception e) {
-        logger.error("Fail to add customized state listener for instance: " + instanceName, e);
+    synchronized (_lastSeenCustomizedStateTypes) {
+      Set<String> lastSeenCustomizedStateTypes = _lastSeenCustomizedStateTypes.get();
+      for (String customizedState : customizedStateTypes) {
```

Comment:
Note that if prefetch is false, you should not expect customizedStateTypes to be filled with data.
The code works here because you are not following the prefetch configuration in the callback handler logic.

Reply:
I made the change. However, I feel we actually do not need the prefetch logic for this event. It's redundant now. Please let me know.